src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+#include <ctype.h>
+#include <dirent.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <sol-log.h>
+#include <sol-mainloop.h>
+#include <sol-str-slice.h>
+#include <sol-util.h>
+#include <sol-util-linux.h>
+#include <sol-vector.h>
+
+#include "sol-iio.h"
+
+struct sol_iio_device {
+    char *trigger_name;
+    sol_iio_reader_cb reader_cb;
+    void *reader_cb_data;
+    struct sol_fd *fd_handler;
+    struct sol_ptr_vector channels;
+    int device_id;
+    int trigger_id;
+    int fd;
+    bool buffer_enabled : 1;
+    bool manual_triggering : 1;
+};
+
+struct sol_iio_channel {
+    struct sol_iio_device *device;
+    char *name;
+    double scale;
+    int index;
+    int offset;
+
+    int storagebits;
+    int bits;
+    int shift;
+    int offset_in_buffer;
+
+    bool little_endian : 1;
+    bool is_signed : 1;
+};
+
+struct sol_iio_buffer {
+    size_t size;
+    uint8_t *data;
+};
+
+#define DEVICE_PATH "/dev/iio:device%d"
+#define SYSFS_DEVICES_PATH "/sys/bus/iio/devices"
+#define BUFFER_ENABLE_DEVICE_PATH "/sys/bus/iio/devices/iio:device%d/buffer/enable"
```

Comment:
Most of these have the same prefix (`"/sys/bus/iio/devices/"`); it could be easier to have a `#define BUS_PREFIX` with that and use it elsewhere. Might be used even for other constants, such as `CHANNEL_RAW_PATH`, `CHANNEL_SCAN_ENABLE_PATH`, and so on. Makes changing these things easier (even changing one define only for testing purposes: say you mock a filesystem with the values and change the root #define to point to that path).